src/witness_generator/machines/block_machine.rs

```diff
@@ -0,0 +1,556 @@
+            match (l, self.evaluate(fixed_data, r)) {
+                (Ok(l), Ok(r)) => match (l.clone() - r).solve_with_bit_constraints(self) {
+                    Ok(result) => results.extend(result),
+                    Err(e) => errors.push(e),
+                },
+                (Err(e), Ok(_)) => errors.push(e.clone()),
+                (Ok(_), Err(e)) => errors.push(e),
+                (Err(e1), Err(e2)) => errors.extend([e1.clone(), e2]),
+            }
+        }
+        if results.is_empty() && !errors.is_empty() {
+            Err(errors.into_iter().reduce(eval_error::combine).unwrap())
+        } else {
+            Ok(results)
+        }
+    }
+
+    fn process_polynomial_identity(
+        &self,
+        fixed_data: &FixedData,
+        identity: &Expression,
+    ) -> EvalResult {
+        let evaluated = self.evaluate(fixed_data, identity)?;
+        evaluated.solve_with_bit_constraints(self).map_err(|e| {
+            let witness_data = WitnessData {
+                fixed_data,
+                data: &self.data,
+                row: self.row,
+            };
+            let formatted = evaluated.format(&witness_data);
+            if let EvalError::ConstraintUnsatisfiable(_) = e {
+                EvalError::ConstraintUnsatisfiable(format!(
+                    "Constraint is invalid ({formatted} != 0)."
+                ))
+            } else {
+                format!("Could not solve expression {formatted} = 0: {e}").into()
+            }
+        })
+    }
+
+    fn process_plookup(
+        &self,
+        fixed_data: &FixedData,
+        fixed_lookup: &mut FixedLookup,
+        identity: &Identity,
+    ) -> EvalResult {
+        if identity.left.selector.is_some() || identity.right.selector.is_some() {
+            unimplemented!("Selectors not yet implemented.");
+        }
+        let left = identity
+            .left
+            .expressions
+            .iter()
+            .map(|e| self.evaluate(fixed_data, e))
+            .collect::<Vec<_>>();
+        if let Some(result) =
+            fixed_lookup.process_plookup(fixed_data, identity.kind, &left, &identity.right)
+        {
+            result
+        } else {
+            Err("Could not find a matching machine for the lookup."
+                .to_string()
+                .into())
+        }
+    }
+
+    fn evaluate(
+        &self,
+        fixed_data: &FixedData,
+        expression: &Expression,
+    ) -> Result<AffineExpression, EvalError> {
+        ExpressionEvaluator::new(SymoblicWitnessEvaluator::new(
+            fixed_data,
+            self.row,
+            WitnessData {
+                fixed_data,
+                data: &self.data,
+                row: self.row,
+            },
+        ))
+        .evaluate(expression)
+    }
+
+    /// Converts a poly ID that migth contain a next offset
+    /// to a regular poly ID plus a row offset.
+    fn convert_next_to_offset(&self, id: usize) -> (usize, u64) {
```

Comment:
Is it correct that the offset is always 0 or 1? If so if would be clearer to use `bool is_next` and cast to u64 on usage.

Reply:
Ah, didn't know you could cast bool to u64 directly.